pkg/git/git.go

```diff
@@ -269,6 +269,16 @@ func updateRepo(repoPath string) (*Repo, error) {
 	}
 
 	// Update the repo
+	if err := command.NewWithWorkDir(
```

Comment:
I usually had an `autostash` in the pull command below but this was not supported by every git version. So I think this is fine, but do we have to `pop` the stash after the `pull`?

Reply:
> […] but do we have to pop the stash after the pull?

For the `krel release-notes` I don't think we should.

We *only* want our update in there, not anything else that happened to be unsaved in repo before we started. We could help that by changing the `git add .` we do later to a git add of just the one file we changed. But even then we get issues. What if the pop doesn't apply cleanly? That means we'd have to fail for reasons unrelated to making the release notes.

We could pop the stash after we've pushed our branch. Or more accurately:

1. stash
2. checkout master
3. pull
4. checkout our desired branch
4. write release notes
5. add release notes file
6. commit
7. push branch
8. checkout the branch we had at he start in step 1
9. pop stash

That would let us do our generation cleanly, but return the repo to how we found it after. Another improvement could be to do a fetch and not a pull, make our branch from wherever we are, and then `git reset --hard` it back to `origin/master`. Such that we never actually touch the local's `master` branch (in case the user has extra commits on it or something, which we would accidentally include in the method above)

Reply:
This goes into the direction of restoring state, which I think is wanted. We could save the currently checked out branch and stash on repo open and provide a method to manually restore that state at the end.

So, I changed my mind and agree that we should checkout the master on open, because if we clone then we’re also on the clean master (same states). 👍

WDYT?

Reply:
> […] We could save the currently checked out branch and stash on repo open and provide a method to manually restore that state at the end.

Something like this?

```
// Ensures that the directory at repoPath has a an up-to-date checkout of `master`
// from the git repository at repoURL. If repoPath is already a git repository, the current
// index and working directory will be stashed and `master` checked out and updated.
//
// The returned reset function can be used to reset the branch, working directory, and
// index to how it was when we started. In the event that we cloned the repository fresh
// the reset function will return the local repository to `master`.
//
// If the given repoURL is not a remote on an existing local repository, it will be added
// using the first available name of `origin`, `upstream`, `kubernetes`, and finally a
// randomly generated name.
func CloneOrOpenRepo(repoPath, repoURL string) (*Repo, func()error, error) {
```

I tried to think of the various things that could go wrong. Are you happy with adding the URL as a remote if it's not there? The alternative is to just fail if repoPath is a git repo but repoURL is not a remote. I should probably bulk out the tests around the cases mentioned too.

> So, I changed my mind and agree that we should checkout the master on open, because if we clone then we’re also on the clean master (same states)

I guess if we wanted to be fully configurable about it, the `CloneOrOpenRepo` function could take in a branch name and would either clone that branch or would pull it. But I don't think there's a use-case yet where we don't want `master` so that might not be needed at this stage.

Reply:
How about:
```
// CloneOrOpenRepo ensures that the directory at repoPath has a an up-to-date checkout 
// of the `master` branch.
//
// If repoPath is already a git repository, the current index and working directory will be
// stashed and `master` checked out and updated.
//
// The checked out branch as well as the saved stash can be restored by calling the
// method `Restore()`, which will have no effect at all if the repository is based on a fresh clone.
//
// If the given repoURL is not a remote on an existing local repository, then this function
// will fail with an error.
func CloneOrOpenRepo(repoPath, repoURL string) (*Repo, error) {
```

Handling the remotes by us would imply more logic and I'm not sure if this is necessary right now. Consumers could also expect that `origin` is the remote which needs to be used, which might be a root cause for errors.

If SSH is being used or not could be automatically detected by the repository URL :+1:

Reply:
> Handling the remotes by us would imply more logic and I'm not sure if this is necessary right now

If the directory at `repoPath` is already a git repository should we at least check that it has `repoURL` as a remote? Or should we ignore `repoURL` completely in that case and instead just pull `master` from whatever the already configured upstream is? (Failing if `master` is either not available locally or if it has no configured upstream.)

> If SSH is being used or not could be automatically detected by the repository URL 

Do we need to do anything different if it is? Don't we just pass the `repoURL` to git and let it figure it out?

> // The checked out branch as well as the saved stash can be restored by calling the
> // method `Restore()`, which will have no effect at all if the repository is based on a
>  // fresh clone.

Two things here. I was expecting we'd return a `func()error` as the reset function, but this docstring seems seems to suggest it's a method on the `Repo` struct instead? I don't really have an objection and it keeps things together somewhat.

The no effect on a fresh clone is interesting. I'd have expected it to return you to `master`. Otherwise you'll end up in a weird state after doing this on the same repo a bunch of times. Take a function that branched and made a PR (like we do for `krel release-notes`).

1. On the first time we execute `CloneOrOpenRepo` will clone the repository afresh, the code branches to `some-branch-1`, pushes it, and calls `Restore()` which does nothing.
2. On the second time `CloneOrOpenRepo` checks `master` back out, pulls it, branches to `some-branch-2` pushes it, and calls `Restore()` which takes us back to `some-branch-1`.
3. On every subsequent time we do the same, always restoring back to `some-branch-1`.

If `Restore()` instead took you to `master` on a clean checkout, we'd end up on `master` all the time.

Reply:
> > Handling the remotes by us would imply more logic and I'm not sure if this is necessary right now
> 
> If the directory at `repoPath` is already a git repository should we at least check that it has `repoURL` as a remote? Or should we ignore `repoURL` completely in that case and instead just pull `master` from whatever the already configured upstream is? (Failing if `master` is either not available locally or if it has no configured upstream.)

Yeah I agree, we could add some basic checks around that. :+1: 

> > If SSH is being used or not could be automatically detected by the repository URL
> 
> Do we need to do anything different if it is? Don't we just pass the `repoURL` to git and let it figure it out?

Yes, you're right here as well :+1: 

> > // The checked out branch as well as the saved stash can be restored by calling the
> > // method `Restore()`, which will have no effect at all if the repository is based on a
> > // fresh clone.
> 
> Two things here. I was expecting we'd return a `func()error` as the reset function, but this docstring seems seems to suggest it's a method on the `Repo` struct instead? I don't really have an objection and it keeps things together somewhat.

Yes, I think we should add it as a method of `Repo`.

> The no effect on a fresh clone is interesting. I'd have expected it to return you to `master`. Otherwise you'll end up in a weird state after doing this on the same repo a bunch of times. Take a function that branched and made a PR (like we do for `krel release-notes`).
>
> 1. On the first time we execute `CloneOrOpenRepo` will clone the repository afresh, the code branches to `some-branch-1`, pushes it, and calls `Restore()` which does nothing.
> 2. On the second time `CloneOrOpenRepo` checks `master` back out, pulls it, branches to `some-branch-2` pushes it, and calls `Restore()` which takes us back to `some-branch-1`.
> 3. On every subsequent time we do the same, always restoring back to `some-branch-1`.
> 
> If `Restore()` instead took you to `master` on a clean checkout, we'd end up on `master` all the time.

Hah yeah, I think we should modify that to return to the master everytime.

Reply:
Been thinking on this some more (apologies, been really busy lately) and I'm not sure if `stash` is a great idea. If the application made another stash in the meantime, then doing a `stash pop` at the end would do the wrong thing.

I wonder if we should actually do a *commit* and soft reset it afterwards? Although handling the index would be a little difficult... I'd like the semantics of the `Restore()` method to be like this:

```
// Restore will return the HEAD, working directory, and index to the state they were in before the application made modifications.
// If HEAD was detached, then it will be returned to the same commit. If HEAD was a branch then HEAD will be returned to this branch, and the branch hard reset to the commit it was on at the time. If HEAD was a tag then the HEAD will be returned to this tag, and the tag reset to the commit it was on at the time.
// Changes to the working directory or index will be restored.
func (r *Repo) Restore() error {
```

But is that doing too much? If we *don't* reset the tag/branch then the stash might not apply cleanly of course.

Reply:
> Been thinking on this some more (apologies, been really busy lately) and I'm not sure if `stash` is a great idea. If the application made another stash in the meantime, then doing a `stash pop` at the end would do the wrong thing.

Do you have a scenario in mind which would fail? I think since the stash is an usual stack, stashing multiple times should work, because the `stash pop` only affects the latest `stash`.

Reply:
> Do you have a scenario in mind which would fail? 

If the rest of the code does something "unusual" with `stash` between opening the repo and calling `Restore()`. For example:

- If they pop the stash without first stashing something then they'll get the changes we were going to restore, and when `Restore()` is called we'll pop whatever happened to be on the stash before we started, or fail if the stash is empty.
- If they push something onto the stash and *don't* pop it we'll pop that instead of what we meant to restore. Which might not apply cleanly.

It could be argued that these are "obvious" things to not do. But then I worry that `Restore()` becomes a leaky abstraction, with a long list of things that you shouldn't do if you want it to work correctly.

More generally the semantics of the `Restore()` function get a little weird when you think about it:

- Some actions we restore, mostly around `HEAD`. But we *don't* want to remove branches we made, reset branches we advanced, or remove tags we made. But what if we make commits on a branch that was `HEAD` before we started? DO we reset that locally as part of `Restore()`.
- What happens if you open a repo, do some work, call `Restore()`, do some more work, and then call `Restore()` a second time. Is that supported?

These could be handled through documentation, but the more and more corner cases I think of the more we might have to document as out of scope. For example:

- What if the program deletes the entire repository?
- What if the program adjusts remotes? What if they adjust the remote for the branch we're going to restore onto?
- What if this is done on a blank repo? (Should `Restore()` try to erase all commits? 😱)
- Should we try to 'fix' the reflog? Or leave it alone?

I can see a few solutions:

1. Document most of these as out of scope, and say that the `Restore()` function is best effort and might fail if you've done something unusual to the repository.
2. Try to handle most of these cases, or at least provide nice error messages.
3. Give up, don't implement `Restore()` and make it the responsibility of the calling code to undo things.
4. Give up the idea of re-using repos and insist that we only do clean checkouts that we can throw away after.

Reply:
Solution nr 4 was the original approach. Since it is an ephemeral fork I don't see why not. Perhaps we could shift our efforts to make sure our local is clean and only act if it is ?

Reply:
Yeah do we want to agree on 3. or 4.? Every krel subcommand should leave the repo in a state which is clean? What means "clean"? I think we could agree on that having no uncommited changes seems to be a factor.

Let's move forward with this PR, I think the improvements are worth to be merged very soon :)

Reply:
Having thought about it some, and discussed a little with @puerco informally, I'm thinking that 4 is the best approach: Use a randomly generated temporary directory name, and delete it afterwards. I think can also save some bandwidth by doing a shallow clone.

I'm thinking of making this explicit too, with a new function like `CleanCloneRepo(repoURL string, shallow bool)` which returns a `Repository` which is guaranteed to be a clean checkout.

For the cleanup we could add a `Cleanup()` method on the `Repository` interface (like `Restore()` as discussed above) but I'm not sure we'd want anything that can operate on a `Repository` to be able to do the cleanup as it's a destructive operation, and repositories from other sources (e.g., the existing `CloneOrOpenRepo()`) don't make sense having such a `Cleanup()` method.

Alternatively we could have a `CleanableRepository` interface implemented like:
```
type CleanableRepository interface {
	Repository
	Cleanup() error
}
```

Reply:
Sounds good, thank you! :)

Reply:
There already is a `Cleanup() error` function, as it happens. So even easier.

I've updated the PR with a rebased implementation.

Reply:
I've also deferred the shallow clone thing for now. I figure we can do that in a future PR.